Modify Basic.doit

mcpl-sympy opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Related to #18837

#### Brief description of what is fixed or changed
`Basic._eval_doit` is introduced, which can be overridden to modify `doit` behavior of subclasses. All subclasses are now expected to define `_eval_doit` method, not `doit`. 
For 'evaluatable' classes (which take `evaluate` option, i.e. `Add`, `Mul`), `Basic._eval_doit_evaluatable` method is prepared.  
`properties` option in introduced to `doit`, which can specify the classes to be 'doit-ed'.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- core
    - `doit` now can take `properties` option, which specifies the classes that are to be affected.
    - `doit` now refers to `_eval_doit` method.
<!-- END RELEASE NOTES -->

JSS95 committed:
Modify Basic.doit

JSS95 committed:
Update doctest and add test

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1624,31 +1624,75 @@ def count_ops(self, visual=None):
+        for evaluatable classes, i.e. Add, Mul, etc.
+        """
+        deep = hints.get('deep', True)
+        properties = hints.get('properties', [])
```
Why is this logic duplicated in `_eval_doit_evaluatable` and `doit`. Surely it only needs to be in the `doit` method.

mcpl-sympy replied:
Oops. Thanks.

JSS95 committed:
Change Atom.doit to Atom._eval_doit

Remove duplicate codes

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1624,31 +1624,73 @@ def count_ops(self, visual=None):
+        eval_doit = self._eval_doit(**hints)
+        if eval_doit is not None:
+            return eval_doit
+        if deep:
```
If deep is true then we should have the args doit before calling `_eval_doit` here.

mcpl-sympy replied:
I don't want to restrict the ability of `_eval_doit` here. Some classes may want to do some processings with unevaluated `self.args` before evaluating themselves.

mcpl-sympy replied:
Oh, on second thought, I think you're right.

JSS95 committed:
Modify _eval_doit

Fixed docstrings and modified every 'doit's to take
options. These doits will later be changed to
_eval_doit.

JSS95 committed:
Fix Atom._eval_doit

JSS95 committed:
Improve code quality

JSS95 committed:
Fix bug in physics.quantum module

JSS95 committed:
Include condition for self.is_Atom in Basic.doit

asmeurer commented on sympy/core/basic.py:
```diff
@@ -1623,32 +1623,70 @@ def count_ops(self, visual=None):
+        integrals, sums and products.
+
+        Arguments will be evaluated recursively, unless some species were
+        excluded via 'properties' or unless the 'deep' hint was set to 'False'.
```
"species" is not a common term used in SymPy (do you mean classes?). Also "properties" is not a very clear parameter name.

mcpl-sympy replied:
Well, I believe the original author ment "classes" indeed. I'll change it.
I borrowed the name "properties" from `Wild`, but have been uneasy with this name as well. Any suggestions?

asmeurer commented:
Not sure about the "properties" flag. Are there examples of that sort of thing in other sympy functions?

oscarbenjamin commented:
I don't think we need the properties flag for now. If we can get a clean model for `doit` and `_eval_doit` then we can think about how to extend that later. If we did have anything like that then it should be in `Basic.doit` rather than `_eval_doit_evaluatable`. I don't see why `_eval_doit_evaluatable` is needed.

oscarbenjamin commented on sympy/integrals/rubi/utility_function.py:
```diff
@@ -444,7 +444,7 @@ def ArcCosh(a):
 
 class Util_Coefficient(Function):
-    def doit(self):
+    def doit(self, **hints):
```
It would be better if this was change to `_eval_doit` rather than modifying the signature of `doit`.

mcpl-sympy replied:
Got it.

JSS95 committed:
Reflect the reviews

mcpl-sympy commented:
@oscarbenjamin
Some classes are evaluatable(e.g. `Derivative`), while some classes are not (e.g. `Matrix`). Currently, how evaluatable classes handle the evaluation varies: `Function` uses `eval`, `Derivative` defines its evaluation in `__new__`, `Add` and `Mul` have `flatten` and `_from_args`, and so on.  

Therefore, I am planning to make these classes follow this one rule: __"Whatever methods they use, make sure that `cls(*args, evaluate=True)` returns evaluated result."__  

With this, my plan A was to make unevaluatable classes take `evaluate` option as well (which will be ignored in this case). With every classes being able to take this option, I would have made `Basic.doit` to be `if not deep: return self.func(*self.args, evaluate=False)`, and make every classes inherit this. With this design, one will define `_eval_doit` in "very" special cases only.
However, many classes do not take keyword arguments at all, which raised error when I changed `Basic.doit` like that. Changing every signature of these classes to solve it seemed to be a bad idea.   

So I moved to plan B: make `self.doit(deep=False)` return `self` if the class does not take `evaluate` option; make it return `self.func(*self.args, evaluate=False)` if it does. In previous commit, I defined `Mul._eval_doit(deep=False)` and `Add._eval_doit(deep=False)` to return this. Then, `doit` on `Add` and `Mul` would return this result. Instead of using same code here and there alike, I moved it to `Basic._eval_doit_evaluatable` and made evaluatable classes use this as `_eval_doit`.

JSS95 committed:
Edit docstring a little bit

JSS95 committed:
Improve code quality

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1623,32 +1623,72 @@ def count_ops(self, visual=None):
 
+    def _eval_doit(self, *args, **hints):
+        """
+        If `deep=True` is passed to `doit` which called this method,
```
If this method receives the args as an argument then I guess it can be a class method and then it can be called from `__new__`

oscarbenjamin replied:
Whether ot not `_eval_doit` is a class method is a key point that needs to be resolved.

mcpl-sympy replied:
Agree. I am also thinking to use the name `eval` (as `Function` does) instead of `_eval_doit`

asmeurer commented:
> @oscarbenjamin
> Some classes are evaluatable(e.g. `Derivative`), while some classes are not (e.g. `Matrix`). Currently, how evaluatable classes handle the evaluation varies: `Function` uses `eval`, `Derivative` defines its evaluation in `__new__`, `Add` and `Mul` have `flatten` and `_from_args`, and so on.
> 
> Therefore, I am planning to make these classes follow this one rule: **"Whatever methods they use, make sure that `cls(*args, evaluate=True)` returns evaluated result."**
> 
> With this, my plan A was to make unevaluatable classes take `evaluate` option as well (which will be ignored in this case). With every classes being able to take this option, I would have made `Basic.doit` to be `if not deep: return self.func(*self.args, evaluate=False)`, and make every classes inherit this. With this design, one will define `_eval_doit` in "very" special cases only.
> However, many classes do not take keyword arguments at all, which raised error when I changed `Basic.doit` like that. Changing every signature of these classes to solve it seemed to be a bad idea.
> 
> So I moved to plan B: make `self.doit(deep=False)` return `self` if the class does not take `evaluate` option; make it return `self.func(*self.args, evaluate=False)` if it does. In previous commit, I defined `Mul._eval_doit(deep=False)` and `Add._eval_doit(deep=False)` to return this. Then, `doit` on `Add` and `Mul` would return this result. Instead of using same code here and there alike, I moved it to `Basic._eval_doit_evaluatable` and made evaluatable classes use this as `_eval_doit`.

This all really underscores the mess that the class design is. If every class were designed like the matrix expressions classes we wouldn't have to make this distinction. It's difficult to change compatibility in that way, though.

oscarbenjamin commented:
I think that the model should be this:

- Evaluation whether automatic in `__new__` or as a result of `doit` is *always* handled by an `_eval_doit` method.
- Some objects evaluate automatically by default so that unless you pass evaluate=False then `_eval_doit` will be called.
- Other objects do not evaluate automatically so that `_eval_doit` is only called if you pass evaluate=True. There objects will default to being created as unevaluated.
- Objects that auto-evaluate can be prevented from evaluating using either evaluate=False or global evaluate. The result is an object that has been "disevaluated".
- Either way these should be equivalent for all Basic instances `obj`:
-- `obj.func(*obj.args, evaluate=True)`
-- `obj.func(*obj.args, evaluate=False).doit(deep=False)`
-- `obj.func(*obj.args).doit(deep=False)`
The key invariant only holds for disevaluated objects:
-- `obj.func(*obj.args) == obj`

I think it's important to move all of the evaluation code out of all of the `__new__` methods so that `Basic.__new__` and `Basic.doit` can use the same method for controlling evaluation however it is triggered.

oscarbenjamin commented:
@mcpl-sympy what's the status of this?

This is not an easy issue to fix and I think it needs some discussion/consensus before implementation. Github indicates that you are a "first time contributor" in which case I have to say that you have not picked an easy issue to work on!

mcpl-sympy commented:
@oscarbenjamin 
Currently I am (or, we are, I should say, since this is the public account for my group) addressing to some other matters, so no progress have been made yet. Still, this issue is on the top priority.

czgdp1807 commented:
@mcpl-sympy Any news on this?

mcpl-sympy commented:
@czgdp1807
Currently working on it.

czgdp1807 commented:
@oscarbenjamin @asmeurer Do you think that the work in this PR can be continued? If yes, then this can be marked as `Please Take Over` PR and explicitly mentioned in the original issue. Thank you.

oscarbenjamin commented:
I think the mechanism here is overly complicated. We should just have doit call `_eval_doit` and not add new arguments like `properties` yet. The `doit` function should just call `doit` (or perhaps `_eval_doit`) on the args and should not use `evaluate=True/False` which is not yet uniformly accepted throughout the codebase and is probably not accepted in downstream code as well.

I suggest something like this:
```python
class Basic:

    def doit(self, **hints):
        # Checking and preprocessing of hints happens here once only
        check(hints)
        deep, opts = process(hints)
        return self._doit(deep, opts)

    def _doit(self, deep, opts):
        # backwards compatibility for subclasses that have defined doit:
        if type(self).doit is not Basic.doit:
            return self.doit(**convert_opts(deep, opts))
        
        args = self.args
        if deep:
            args = [a._doit(deep, opts) for a in args]

        if args != self.args:
            obj = self.func(*args)
        else:
            obj = self

        result = obj._eval_doit()
        if result is not None:
            obj = result

        return obj

    def _eval_doit(self):
        pass
```

Then the idea is that all subclasses around the codebase that have defined doit methods should be changed to use `_eval_doit`.

Key issues:

1. Need to ensure backwards compatibility for downstream code that has defined doit methods
2. Should `_eval_doit` be a class method so that it can be used in `__new__` without calling `Basic.__new__`?
3. Should we allow custom options in `doit` to be handled in `_eval_doit` or should we define options in a systematic way that are handled by `_doit` or should we allow both?

For point 1 I think what I show above should be okay.

For point 2 I would be interested to see a performance analysis of the code of always calling `Basic.__new__(*args)._eval_doit()`. Having `_eval_doit` as an instance method would simplify a lot of the logic in many cases as it means that instance methods and properties are available. Skipping `Basic.__new__` is a widely used optimisation but I have no idea how beneficial it actually is. Perhaps there are a few key classes (e.g. `Add`, `Mul`, `Pow`) where this is worth doing in which case the mechanism for `eval/doit` could be different just for those classes but all others would need to call `Basic.__new__` and could have `_eval_doit` as an instance method.

For point 3 by systematic options I mean something like:
```
    obj.doit(exclude={Integral})
```
There could be options for deciding which objects to call `_eval_doit` on and those could be handled in `_doit`. The custom options would be something specific to a certain class like
```
    obj.doit(integrate_method='merijerg')
```
That kind of option would need to be passed to e.g. `Integral._eval_doit`.
